### PR TITLE
#34 added test to prove Xembeler is not thread-safe yet

### PR DESCRIPTION
**Pull request by apocarteres**

### CODE REVIEW

**Comment by davvd** [comment]
Thanks, I'll find someone to review this PR soon

**Comment by davvd** [comment]
@yegor256 review this one, please

File: src/test/java/org/xembly/XemblerTest.java [hunk 1]
```diff
@@ -177,4 +185,67 @@ public void replacesRootNode() throws Exception {
...
+                .newDocumentBuilder().newDocument();
+        final Xembler xembler = new Xembler(
+            new Directives(
+                "ADD 'root';ADDIF 'blow';"
+            )
+        );
+        final int capacity = 10000;
+        final Collection<Callable<Document>> tasks =
+            new ArrayList<Callable<Document>>(
+                    capacity
```
Comment (yegor256) [inline src/test/java/org/xembly/XemblerTest.java]: why 8 indentation spaces?

File: src/test/java/org/xembly/XemblerTest.java [hunk 2]
```diff
@@ -177,4 +185,67 @@ public void replacesRootNode() throws Exception {
...
+            );
+        for (int idx = 0; idx < capacity; ++idx) {
+            final Callable<Document> callable = XemblerTest.buildCallable(
+                    xembler, dom
```
Comment (yegor256) [inline src/test/java/org/xembly/XemblerTest.java]: again, why eight?

File: src/test/java/org/xembly/XemblerTest.java [hunk 3]
```diff
@@ -177,4 +185,67 @@ public void replacesRootNode() throws Exception {
...
+            );
+            tasks.add(callable);
+        }
+        final List<Future<Document>> futures = service.invokeAll(tasks);
+        for (final Future<Document> future : futures) {
+            final Document document = future.get();
+            MatcherAssert.assertThat(
+                XhtmlMatchers.xhtml(document),
+                XhtmlMatchers.hasXPaths(
+                        "/root[count(blow) = 1]"
```
Comment (yegor256) [inline src/test/java/org/xembly/XemblerTest.java]: same here

File: src/test/java/org/xembly/XemblerTest.java [hunk 4]
```diff
@@ -177,4 +185,67 @@ public void replacesRootNode() throws Exception {
...
+    }
+
+    /**
+     * Builds up callable object.
+     * @param xembler Xembler's instance
+     * @param document DOM object
+     * @return Callable object
+     */
+    private static Callable<Document> buildCallable(
+            final Xembler xembler, final Document document
```
Comment (yegor256) [inline src/test/java/org/xembly/XemblerTest.java]: and here

File: src/test/java/org/xembly/XemblerTest.java [hunk 5]
```diff
@@ -177,4 +185,67 @@ public void replacesRootNode() throws Exception {
...
+        service.shutdown();
+    }
+
+    /**
+     * Builds up callable object.
+     * @param xembler Xembler's instance
+     * @param document DOM object
+     * @return Callable object
+     */
+    private static Callable<Document> buildCallable(
```
Comment (yegor256) [inline src/test/java/org/xembly/XemblerTest.java]: I would name it `callable`, see http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#method-names

**Comment by yegor256** [comment]
@apocarteres looks good, but I didn't get why sometimes you're using 4 spaces and sometimes 8.. what is the logic? We're using 4, always

**Comment by apocarteres** [comment]
yeah no logic then, let me fix IDE codestyle. should i fix indents and re-initiatte PR?

**Comment by yegor256** [comment]
@apocarteres even though the test looks correct (syntax/grammar), I don't understand its logic. You're trying to modify the same (!) DOM from 10000 different threads. All of that modifications will succeed only if their directives will be `ADDIF 'root'; ADDIF 'blow';`. Pay attention, the first one should be `ADDIF`. If you use `ADD`, they all will try to create a new (!) root element. This will lead to an error in any case. Either you do them in parallel, or consequently in one thread.

So, the correct logic would be to start with a `<root>` element and run `ADDIF 'root'; ADDIF 'blow';` from multiple threads. If you run them in one thread, they all will succeed (try it).

**Comment by apocarteres** [comment]
thank you for review. Yes, test ought to be amended, but it still proves an issue. I followed your advice and there is <code><root></code> node which initialized out of concurrent environment, and directives <code>"ADDIF 'blow';REMOVE;ADDIF 'blow';"</code> which get applied through multithreaded invocation. pls check it.

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by davvd** [comment]
@yegor256 Thanks a lot, I just topped your account for **16 mins**, transaction ID `000-96986aea`. +16 added to your rating, at the moment it is: [+4334](http://www.netbout.com/b/35008?open=rating)

**Comment by davvd** [comment]
@rultor deploy

**Comment by yegor256** [comment]
@rultor deploy

**Pull request merged**